internal/backend/workloadproxy/reconciler.go

```diff
@@ -6,83 +6,68 @@
 package workloadproxy
 
 import (
-	"errors"
+	"context"
 	"fmt"
 	"net"
 	"net/http"
 	"net/http/httputil"
-	"net/url"
+	"slices"
+	"strings"
 	"sync"
 	"time"
 
-	"github.com/akutz/memconn"
 	"github.com/cosi-project/runtime/pkg/resource"
 	"github.com/hashicorp/go-multierror"
-	"github.com/siderolabs/go-loadbalancer/loadbalancer"
+	"github.com/siderolabs/gen/xiter"
 	"github.com/siderolabs/go-loadbalancer/upstream"
-	"github.com/siderolabs/tcpproxy"
 	"go.uber.org/zap"
 	"go.uber.org/zap/zapcore"
+	"golang.org/x/net/http2"
 )
 
-type lbStatus struct {
-	lb *loadbalancer.TCP
-}
-
-// Close closes the load balancer and waits for it to terminate.
-//
-// It should be called only if it was successfully started, otherwise it blocks indefinitely on the Wait call.
-func (lbSts *lbStatus) Close() error {
-	var lb *loadbalancer.TCP
-
-	lb, lbSts.lb = lbSts.lb, nil
-
-	if lb == nil {
-		return nil
-	}
-
-	if err := lb.Close(); err != nil {
-		return fmt.Errorf("failed to close LB: %w", err)
-	}
-
-	if err := lb.Wait(); err != nil && !errors.Is(err, net.ErrClosed) {
-		return fmt.Errorf("failed to wait for load balancer: %w", err)
-	}
-
-	return nil
-}
-
 // Reconciler reconciles the load balancers for a cluster.
+//
+//nolint:govet
 type Reconciler struct {
-	clusterToAliasToLBStatus map[resource.ID]map[string]*lbStatus
-	aliasToCluster           map[string]resource.ID
-
-	connProvider *memconn.Provider
-	logger       *zap.Logger
-	lbLogger     *zap.Logger
-	logLevel     zapcore.Level
-	mu           sync.Mutex
+	logger    *zap.Logger
+	logLevel  zapcore.Level
+	transport *http.Transport
+
+	mu                      sync.Mutex
+	clusterToAliasToBackend map[resource.ID]map[string]*upstream.List[*backend]
+	aliasToCluster          map[string]resource.ID
 }
 
 // NewReconciler creates a new Reconciler.
 func NewReconciler(logger *zap.Logger, logLevel zapcore.Level) *Reconciler {
-	// use an in-memory transport for the connections to the load balancer
-	provider := &memconn.Provider{}
-
-	provider.MapNetwork("tcp", "memu")
-
 	if logger == nil {
 		logger = zap.NewNop()
 	}
 
-	return &Reconciler{
-		clusterToAliasToLBStatus: map[resource.ID]map[string]*lbStatus{},
-		aliasToCluster:           map[string]resource.ID{},
-		connProvider:             provider,
-		logger:                   logger,
-		lbLogger:                 logger.WithOptions(zap.IncreaseLevel(zapcore.ErrorLevel)),
-		logLevel:                 logLevel,
+	trans := &http.Transport{
+		Proxy:                 http.ProxyFromEnvironment,
+		ForceAttemptHTTP2:     true,
+		MaxIdleConns:          100,
```

Comment:
we need to consider idle connections in a specific way, as the host would be service name

Reply:
there's also stuff like MaxConnsPerHost